server/image_pull.go

```diff
@@ -67,11 +68,20 @@ func (s *Server) PullImage(ctx context.Context, req *pb.PullImageRequest) (*pb.P
 		}
 
 		// let's be smart, docker doesn't repull if image already exists.
-		_, err = s.StorageImageServer().ImageStatus(s.ImageContext(), img)
+		var storedImage *storage.ImageResult
+		storedImage, err = s.StorageImageServer().ImageStatus(s.ImageContext(), img)
 		if err == nil {
-			logrus.Debugf("image %s already in store, skipping pull", img)
-			pulled = img
-			break
+			tmpImg, err := s.StorageImageServer().PrepareImage(s.ImageContext(), img, options)
+			if err != nil {
+				logrus.Debugf("error preparing image, will skip pull: %v", err)
+			}
+			tmpImgConfigDigest := tmpImg.ConfigInfo().Digest
+			if tmpImgConfigDigest.String() == storedImage.ConfigDigest.String() {
```

Comment:
For a schema 1 image, the `ConfigInfo().Digest` value will be empty, so we'll be comparing "" to "" here.

Reply:
Lemme special case that then, since schema1 always results in a pull (Docker always repull with schema1).

Reply:
addressed, PTAL

Reply:
We can probably do slightly better than that.  https://github.com/containers/image/pull/305 adds an ID calculation method to the Manifest type that @mtrmac introduced.  It expects to be handed the image's non-empty diffIDs to calculate the proper ID, but if the goal is only to check for equivalence, you could hand it the list of digests from its own `LayerInfos()` list.

Reply:
> but if the goal is only to check for equivalence, you could hand it the list of digests from its own LayerInfos() list.

That won’t notice if the second image differs only by the adding some config modifications (`ENV`, `RUN`, etc.) on top of the first image.

Reply:
There would be an extra History item in the manifest, wouldn't there?

Reply:
Hmm, right, those changes would be discarded when we parse that field as a SchemaV1Compatibility.

Reply:
`emptyLayer`/`throwaway` history items do not show up in `LayerInfos`.